src/platform/qt/library/LibraryModel.cpp

```diff
@@ -0,0 +1,424 @@
+	}
+	return createIndex(child.internalId() - 1, 0, quintptr(0));
+}
+
+int LibraryModel::columnCount(const QModelIndex& parent) const {
+	if (!parent.isValid() || (parent.column() == 0 && !parent.parent().isValid())) {
+		return MAX_COLUMN + 1;
+	}
+	return 0;
+}
+
+int LibraryModel::rowCount(const QModelIndex& parent) const {
+	if (parent.isValid()) {
+		if (m_treeMode) {
+			if (parent.row() < 0 || parent.row() >= m_pathOrder.size() || parent.column() != 0) {
+				return 0;
+			}
+			return m_pathIndex[m_pathOrder[parent.row()]].size();
+		}
+		return 0;
+	}
+	if (m_treeMode) {
+		return m_pathOrder.size();
+	}
+	return m_games.size();
+}
+
+QVariant LibraryModel::folderData(const QModelIndex& index, int role) const
+{
+	// Precondition: index and role must have already been validated
+	if (role == Qt::DecorationRole) {
+		return qApp->style()->standardIcon(QStyle::SP_DirOpenIcon);
+	}
+	if (role == FullPathRole || (index.column() == COL_LOCATION && role != Qt::DisplayRole)) {
+		return m_pathOrder[index.row()];
+	}
+	if (index.column() == COL_NAME) {
+		QString path = m_pathOrder[index.row()];
+		return path.section('/', -1);
+	}
+	return QVariant();
+}
+
+QVariant LibraryModel::data(const QModelIndex& index, int role) const {
+	if (role != Qt::DisplayRole &&
+			role != Qt::EditRole &&
+			role != Qt::ToolTipRole &&
+			role != Qt::DecorationRole &&
+			role != Qt::TextAlignmentRole &&
+			role != FullPathRole) {
+		return QVariant();
+	}
+	if (!checkIndex(index)) {
+		return QVariant();
+	}
+	if (role == Qt::ToolTipRole && index.column() > COL_LOCATION) {
+		return QVariant();
+	}
+	if (role == Qt::DecorationRole && index.column() != COL_NAME) {
+		return QVariant();
+	}
+	if (role == Qt::TextAlignmentRole) {
+		return index.column() == COL_SIZE ? (int)(Qt::AlignTrailing | Qt::AlignVCenter) : (int)(Qt::AlignLeading | Qt::AlignVCenter);
+	}
+	const LibraryEntry* entry = nullptr;
+	if (m_treeMode) {
+		if (!index.parent().isValid()) {
+			return folderData(index, role);
+		}
+		QString path = m_pathOrder[index.parent().row()];
+		entry = m_pathIndex[path][index.row()];
+	} else if (!index.parent().isValid() && index.row() < m_games.size()) {
+			entry = &m_games[index.row()];
+		}
+		if (entry) {
+			if (role == FullPathRole) {
+				return entry->fullpath;
+			}
+			switch (index.column()) {
+				case COL_NAME:
+					if (role == Qt::DecorationRole) {
+						return m_icons.value(entry->platform, qApp->style()->standardIcon(QStyle::SP_FileIcon));
+					}
+					return entry->displayTitle(m_showFilename);
+				case COL_LOCATION:
+					return QDir::toNativeSeparators(entry->base);
+				case COL_PLATFORM:
+					return nicePlatformFormat(entry->platform);
+				case COL_SIZE:
+					return (role == Qt::DisplayRole) ? QVariant(niceSizeFormat(entry->filesize)) : QVariant(int(entry->filesize));
+				case COL_CRC32:
+					return (role == Qt::DisplayRole) ? QVariant(QStringLiteral("%0").arg(entry->crc32, 8, 16, QChar('0'))) : QVariant(entry->crc32);
```

Comment:
Cases don't get indented.

Reply:
*facepalm* I don't indent them in my typical style, either, but I didn't notice that the autoindent changed it when I was fixing spaces-vs-tabs. ^^()